### PR TITLE
8313138: Scrollbar Keyboard enhancement

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Adding alt-ctrl-LEFT/RIGHT/UP/DOWN (option-command-LEFT/RIGHT/UP/DOWN) key bindings to

- ListView
- TreeView
- TableView
- TreeTableView

to support keyboard-only horizontal and vertical scrolling.  The main reason for the change is to improve accessibility.

**NOTE: For controls in right-to-left orientation, the direction of horizontal scrolling is reversed.**

As far as I can tell, these key combinations do not interfere with editing.

The proposed solution can be further optimized by adding a public method to the VirtualFlow class, something like

```
public void horizontalUnitScroll(boolean right);
public void verticalUnitScroll(boolean down);
```

Q: Does this change require a CSR to explain the change in the controls' behavior?  We don't yet have the key bindings documented in /doc-files/behavior

Note:
Jenkins headful test passed on all mac configurations, failed on all linux configurations (master branch failed also, so it is test issue), while windows configuration is not yet available.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8329336](https://bugs.openjdk.org/browse/JDK-8329336) to be approved
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8313138](https://bugs.openjdk.org/browse/JDK-8313138): Scrollbar Keyboard enhancement (**Enhancement** - P3)
 * [JDK-8329336](https://bugs.openjdk.org/browse/JDK-8329336): Scrollbar Keyboard enhancement (**CSR**)

### Reviewers
 * [Alexander Zuev](https://openjdk.org/census#kizune) (@azuev-java - Author) 🔄 Re-review required (review applies to [5bae5e7a](https://git.openjdk.org/jfx/pull/1393/files/5bae5e7a1725006fd8a9cb32993c67ce828b893b))
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1393/head:pull/1393` \
`$ git checkout pull/1393`

Update a local copy of the PR: \
`$ git checkout pull/1393` \
`$ git pull https://git.openjdk.org/jfx.git pull/1393/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1393`

View PR using the GUI difftool: \
`$ git pr show -t 1393`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1393.diff">https://git.openjdk.org/jfx/pull/1393.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1393#issuecomment-2021071897)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
Since this is a noticeable (end-user-visible) behavioral change, I think a CSR would be in order (it need not be very detailed).

We don't currently document keyboard shortcuts -- either for individual controls or for focus traversal -- although that would be a good doc enhancement (for a later time).

Reviewers: @arapte @azuev-java

/reviewers 2
/csr needed

**Comment by andy-goryachev-oracle** [comment]
@azuev-java could you please take a look at this from accessibility perspective?

**Review by azuev-java — APPROVED** [review]
Looks good.

**Comment by andy-goryachev-oracle** [comment]
@azuev-java does the choice of key combination make sense?

**Comment by kevinrushforth** [comment]
A couple questions about this:

1. What are the keyboard shortcuts for vertical scrolling? I would expect them to be something like alt-ctrl-UP/DOWN but that doesn't seem to be the case. In fact I don't see a way to even do vertical scrolling (PAGE UP/PAGE DOWN is not the same as vertical scrolling), but maybe I'm missing something.
2. What do other apps or toolkits provide for horizontal / vertical scrolling?

**Comment by andy-goryachev-oracle** [comment]
1. Vertical scrolling is not addressed in this PR (the ticket *is* about horizontal scrolling).  I suppose the existing selection/navigation key bindings are sufficient.

2. In most cases, LEFT/RIGHT arrows; I haven't encountered any special keys for horizontal navigation.  We could have used that for ListView case, but it is clearly insufficient for any table-like control.  A multi-key chord was therefore selected that a) works for all 4 cases and b) does not interfere with editing.

**Comment by andy-goryachev-oracle** [comment]
@aghaisas and @arapte could you also review this please?

**Comment by victordyakov** [comment]
@azuev-java please review

**Comment by azuev-java** [comment]
> @azuev-java does the choice of key combination make sense?

Perfectly fine. When Voice Over is active it will override any defined key combination but with Voice Over we do not really care about the scrolling for better readability - and if there is any element that can be selected with VO cursor it will do scrolling by itself.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/ListViewSkin.java [hunk 1]
```diff
@@ -678,4 +678,16 @@ private int onScrollPageUp(boolean isFocusDriven) {
         flow.scrollTo(firstVisibleCell);
         return newSelectionIndex;
     }
+
+    private void horizontalUnitScroll(boolean right) {
+        if (getSkinnable().getEffectiveNodeOrientation() == NodeOrientation.RIGHT_TO_LEFT) {
+            right = !right;
+        }
+        ScrollBar sb = flow.getHbar();
```
Comment (eduardsdv) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/ListViewSkin.java:686]: The ListView has the orientation property ``ListView.setOrientation(Orientation)``.
I think it should be
`````java
ScrollBar sb = flow.isVertical() ? flow.getHbar() : flow.getVbar();
`````
and the name of the method something like ``crossUnitScroll(boolean)``.

**Comment by eduardsdv** [comment]
What about the ``GridView``? It also uses the ``VirtualFlow``.
Should it also support this kind of scrolling?

**Comment by andy-goryachev-oracle** [comment]
> What about the `GridView`

What's a GridView?  ControlsFX is not a part of OpenJFX...

**Comment by eduardsdv** [comment]
> > What about the `GridView`
> 
> What's a GridView? ControlsFX is not a part of OpenJFX...

Oh, sorry, my mistake. You are right. It is part of the ControlsFX.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/ListViewSkin.java (hunk 1 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/ListViewSkin.java:686]: I am afraid this suggestion is incorrect: the new key bindings affect horizontal scroll bar regardless of the orientation.

The new behavior can be checked with the Monkey Tester
https://github.com/andy-goryachev-oracle/MonkeyTest

**Comment by andy-goryachev-oracle** [comment]
... but nothing should prevent the maintainers of ControlsFx from adding this functionality to the GridView.

**Comment by andy-goryachev-oracle** [comment]
As @eduardsdv mentioned in
https://github.com/openjdk/jfx/pull/1326#issuecomment-2096666336

and @kevinrushforth asked about earlier, perhaps we should add similar capability for vertical scrolling using alt-ctrl-Up/DOWN (option-command-UP/DOWN) in the same PR rather than separately.

**Comment by arapte** [comment]
Tested a few samples. It looks good.
I need sometime (a day or two) to complete code review.

**Review by arapte — APPROVED** [review]
LGTM...

**Comment by kevinrushforth** [comment]
As a follow-up enhancement, we might consider extending this to TextArea and/or ScrollPane.

**Review by kevinrushforth — APPROVED** [review]
Looks good. I left one comment (on a comment line) that applies to all 4 controls. I'll reapprove if you make the suggested change.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java [hunk 2]
```diff
@@ -912,4 +929,28 @@ private void discontinuousSelectAllToLastRow() {
 
         if (onMoveToLastCell != null) onMoveToLastCell.run();
     }
+
+    // TODO not necessary with the new InputMap V2
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java]: removed.

**Review by kevinrushforth — APPROVED** [review]

**Review by arapte — APPROVED** [review]

**Pull request closed without merging**